apps/service_providers/llm_service/prompt_context.py

```diff
@@ -137,3 +139,70 @@ def __eq__(self, other):
 
 
 EMPTY = SafeAccessWrapper(None)
+
+
+class ParticipantDataProxy:
+    """Allows multiple access without needing to re-fetch from the DB"""
+
+    def __init__(self, experiment_session):
+        self.session = experiment_session
+        self._participant_data = None
+        self._scheduled_messages = None
+
+    @classmethod
+    def from_state(cls, pipeline_state) -> Self:
+        # using `.get` here for the sake of tests. In practice the session should always be present
+        return cls(pipeline_state.get("experiment_session"))
+
+    def _get_db_object(self):
+        if not self._participant_data:
+            self._participant_data, _ = ParticipantData.objects.get_or_create(
+                participant_id=self.session.participant_id,
+                experiment_id=self.session.experiment_id,
+                team_id=self.session.team_id,
+            )
+        return self._participant_data
+
+    def get(self):
+        data = self._get_db_object().data
+        return self.session.participant.global_data | data
+
+    def set(self, data):
+        participant_data = self._get_db_object()
+        participant_data.data = data
+        participant_data.save(update_fields=["data"])
+
+        self.session.participant.update_name_from_data(data)
+
+    def get_schedules(self):
+        """
+        Returns all active scheduled messages for the participant in the current experiment session.
+        """
+        if self._scheduled_messages is not None:
+            return self._scheduled_messages
+
+        from apps.events.models import ScheduledMessage
+
+        experiment = self.session.experiment_id
+        participant = self.session.participant_id
+        team = self.session.experiment.team
+        messages = (
+            ScheduledMessage.objects.filter(
+                experiment_id=experiment,
+                participant_id=participant,
+                team=team,
+                is_complete=False,
+                cancelled_at=None,
+            )
+            .select_related("action")
+            .order_by("created_at")
+        )
+        scheduled_messages = []
+        for message in messages:
+            scheduled_messages.append(message.as_dict())
+        return scheduled_messages
```

Comment:
This should be inverted so that the cache attribute is set if it is None:
```suggestion
        if self._scheduled_messages is None:
            from apps.events.models import ScheduledMessage
            experiment = self.session.experiment_id
            participant = self.session.participant_id
            team = self.session.experiment.team
            messages = (
                ScheduledMessage.objects.filter(
                    experiment_id=experiment,
                    participant_id=participant,
                    team=team,
                    is_complete=False,
                    cancelled_at=None,
                )
                .select_related("action")
                .order_by("created_at")
            )
            self._scheduled_messages = [message.as_dict() for message in messages]
        return self._scheduled_messages
```